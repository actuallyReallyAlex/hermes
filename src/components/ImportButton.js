import React, { Fragment, createRef } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { replaceShip } from '../redux/actions/ship'
import { replaceUI } from '../redux/actions/ui'
import { replaceUser } from '../redux/actions/user'
import { replaceWorld } from '../redux/actions/world'
import Icon from '@material-ui/core/Icon'

/**
 * Allows the user to import their game from an exported file.
 */
const ImportButton = ({ handleImportGame }) => {
  const inputRef = createRef()
  const handleImportGameClick = () => {
    if (window.Cypress) {
      localStorage.setItem('importedGame', true)
      return
    }
    inputRef.current.click()
  }

  const handleFileAdded = e => {
    const reader = new FileReader()

    reader.onload = e => handleImportGame(JSON.parse(e.target.result))

    reader.onerror = err => {
      console.error(err)
      alert('An error occured while importing the file.')
    }

    reader.readAsText(e.target.files[0])
  }

  return (
    <Fragment>
      <Icon className="fas fa-download" onClick={handleImportGameClick} />
      <input
        accept=".json"
        data-testid="input-import"
        onChange={handleFileAdded}
        ref={inputRef}
        style={{ display: 'none' }}
        type="file"
      />
    </Fragment>
  )
}

ImportButton.propTypes = {
  handleImportGame: PropTypes.func.isRequired
}

const mapDispatchToProps = dispatch => ({
  handleImportGame: ({ ship, ui, user, world }) => {
    dispatch(replaceShip(ship))
    dispatch(replaceUI(ui))
    dispatch(replaceUser(user))
    dispatch(replaceWorld(world))
  }
})

export default connect(
  null,
  mapDispatchToProps
)(ImportButton)
