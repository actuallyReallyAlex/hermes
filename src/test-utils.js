import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { defaultState } from './fixtures'
import { render } from '@testing-library/react'

const mockThunks = ({ dispatch, getState }) => next => action => {
  const mockedAction = () => ({ type: 'MOCKED ACTION' })

  let returnValue

  if (typeof action === 'function' || action === undefined) {
    returnValue = mockedAction
  } else {
    returnValue = next(action)
  }

  return returnValue
}

const middlewares = [mockThunks]

export const mockStore = configureStore(middlewares)

export const customRender = options => {
  const { component, state, props } = options

  const Component = component

  return render(
    <Provider store={mockStore(state ? state : defaultState)}>
      <Component {...props} />
    </Provider>
  )
}
