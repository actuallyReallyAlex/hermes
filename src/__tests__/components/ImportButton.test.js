import { customRender } from '../../test-utils'
import ImportButton from '../../components/ImportButton'
import { fireEvent } from '@testing-library/dom'

describe('<ImportButton />', () => {
  it('Should render the <ImportButton /> component.', () => {
    const container = customRender({ component: ImportButton })
    expect(container.asFragment()).toMatchSnapshot()
  })

  it('Should handle clicking the import button.', () => {
    const { getByText } = customRender({ component: ImportButton })
    const button = getByText('Import Game')
    fireEvent.click(button)
  })

  it('Should handle when a file is selected.', () => {
    const { getByTestId } = customRender({ component: ImportButton })
    const input = getByTestId('input-import')
    const blob = new Blob(['Mock File'])
    fireEvent.change(input, { target: { files: [blob] } })
  })
})
