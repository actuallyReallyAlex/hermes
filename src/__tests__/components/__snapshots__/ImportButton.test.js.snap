// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportButton /> Should render the <ImportButton /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 fWrrBh"
  >
    <button
      class="StyledButton-sc-323bzc-0 jfbKsv"
      type="button"
    >
      Import Game
    </button>
    <input
      accept=".json"
      data-testid="input-import"
      style="display: none;"
      type="file"
    />
  </div>
</DocumentFragment>
`;
